openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py

```diff
@@ -0,0 +1,594 @@
+        "plus_report_line_ids",
+        "plus_report_expression_ids",
+    ),
+]
+_models_renames = [
+    ("account.tax.report", "account.report"),
+    ("account.tax.carryover.line", "account.report.external.value"),
+    ("account.tax.report.line", "account.report.line"),
+]
+_tables_renames = [
+    ("account_tax_report", "account_report"),
+    ("account_tax_carryover_line", "account_report_external_value"),
+    ("account_tax_report_line", "account_report_line"),
+]
+
+
+def _fast_fill_account_account_type(env, model, table):
+    if not openupgrade.column_exists(env.cr, table, "account_type"):
+        openupgrade.add_fields(
+            env,
+            [
+                (
+                    "account_type",
+                    model,
+                    table,
+                    "selection",
+                    False,
+                    "account",
+                )
+            ],
+        )
+    data_query = """SELECT res_id FROM ir_model_data
+        WHERE module='%s' AND name='%s'"""
+    for (account_type_old, account_type_new) in ACCOUNT_TYPE:
+        get_data_query = data_query % tuple(account_type_old.split("."))
+        query = f"""
+            WITH account_type_old AS (
+                SELECT id AS account_id
+                FROM account_account
+                WHERE user_type_id IN (
+                    {get_data_query}
+                )
+            )
+            UPDATE {table}
+            SET account_type = '{account_type_new}'
+            WHERE id IN (SELECT account_id FROM account_type_old)
+        """
+        openupgrade.logged_query(env.cr, query)
+
+
+def _delete_sql_constraints(env):
+    # Delete constraints to recreate it
+    openupgrade.delete_sql_constraint_safely(
+        env, "account", "account_journal", "code_company_uniq"
+    )
+    openupgrade.delete_sql_constraint_safely(
+        env, "account", "account_move_line", "check_accountable_required_fields"
+    )
+    openupgrade.delete_sql_constraint_safely(
+        env, "account", "account_move_line", "check_amount_currency_balance_sign"
+    )
+    openupgrade.delete_sql_constraint_safely(
+        env, "account", "account_move_line", "check_credit_debit"
+    )
+
+
+def _account_analytic_line_fast_fill_journal_id(env):
+    if not openupgrade.column_exists(env.cr, "account_analytic_line", "journal_id"):
+        openupgrade.add_fields(
+            env,
+            [
+                (
+                    "journal_id",
+                    "account.analytic.line",
+                    "account_analytic_line",
+                    "many2one",
+                    False,
+                    "account",
+                )
+            ],
+        )
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_analytic_line AS aal
+        SET journal_id = (
+            SELECT journal_id
+                FROM account_move_line AS aml
+                WHERE aml.id = aal.move_line_id
+            LIMIT 1
```

Comment:
@duong77476 như này cho đơn giản nhé
```suggestion
        UPDATE account_analytic_line aal
            SET journal_id = aml.journal_id
        FROM account_move_line aml
        WHERE aml.id = aal.move_line_id
```

Reply:
Đoạn này trc e code quên chưa sửa, e sẽ sửa lại